software-stack/lab: Explain syscalls (#153)
open-education-hub/operating-systems: content/chapters/software-stack/lab/content/basic-syscall.md

To get a better grasp on how the software stack works, let's do a bottom-up approach:
we build and run different programs, that start of by using the system call API (the lowest layer in the software stack) and progressively use higher layers.

### System Calls Explained

A system call, or syscall for short, is a method used by applications to communicate with the operating system's kernel.

> The need for syscalls is tied to the modern operating systems model of conceptually separate into kernel space and user space. The kernel space is responsible for managing the hardware resources and provide an interface for the user space applications to request information from the hardware. The user space is where you are running your applications and processes.

Reviewer: We should specify what are those hardware resources: CPU, memory, and I/O devices. Also, better than saying "request information" we could add some examples to highlight what a syscall enable the user to do:
- brk() is used to allocate memory
- getpid() ...
- open() ...

Reviewer 2: +1

> The need for syscalls is tied to the modern operating systems model of conceptually separate into kernel space and user space. The kernel space is responsible for managing the hardware resources and provide an interface for the user space applications to request information from the hardware. The user space is where you are running your applications and processes.

Reviewer 2: ```suggestion
The need for syscalls is tied to the modern operating systems model of conceptually separating into kernel space and user space.
The kernel space manages the hardware resources and provides an interface for the user space applications to interact with the hardware.
The user space is where you are running your applications and processes.
```

To get a better grasp on how the software stack works, let's do a bottom-up approach:
we build and run different programs, that start of by using the system call API (the lowest layer in the software stack) and progressively use higher layers.

### System Calls Explained

A system call, or syscall for short, is a method used by applications to communicate with the operating system's kernel.

The need for syscalls is tied to the modern operating systems model of conceptually separating into kernel space and user space.

The kernel space manages the hardware resources such as CPU, I/O devices, disk or memory.
Moreover, the kernel also provides an interface for the user space applications to interact with the hardware.

The user space is where you are running your applications and processes.
From the user space, we cannot directly access the hardware or perform privileged operations.
You need to use syscalls to perform privileged operations such as accessing the hardware.

Below, you can see some examples of system calls and what resource they request from the kernel:

> - `brk()` is used to allocate memory
> - `open()` is used to access the file system and open a specific file
> - `write()` is used to access the file system and modify the contents of a specific file

Reviewer 2: ```suggestion
- `brk()` is used to allocate memory

- `open()` is used to access the file system and open a specific file

- `write()` is used to access the file system and modify the contents of a specific file
```
IDK why markdownlint doesn't complain about this, but it should. If you have lists that span multiple sentences, they're difficult to read unless you leave some blank lines between list items.

Reviewer: It only complains about heterogeneous lists, e.g.: tests that have `+` and `-`.  Though, I think the default line spacing in lists is enough

Author: It seems to me a bit too much to add these newlines. But it is not a strong opinion. I changed it as requested.

> <img src="../media/syscall-explained.svg" alt="System Calls Explained" width="40%" />

Reviewer 2: Don't use inline HTML because it messes up the rendered website (https://github.com/open-education-hub/operating-systems/pull/108#issue-1538121478). Make the image smaller instead.

Reviewer: Resizing is optional, from what I have tested Docusaurus resizes images if the size is below a threshold. We should only focus on the content right now.

Author: Done
